### ISSUE TITLE
Support import / export of external HLB modules, paving the way for HLB libraries

### ISSUE THREAD

**Issue by hinshun** [openllb/hlb#20]
There's two main use cases for libraries:
1. Sharing common hlb functions, like a canonical function for building go code, or running npm install, etc.
2. Sharing a hlb function wrapper for an external frontend.

Either way, we want to avoid copy & pasting that happens in Dockerfiles, and since we have function arguments, we can have better reusability than `ONBUILD` instructions.

Proposal
---

## HLB CLI

HLB is a file-scoped language, so each file represents a module. We introduce a subcommand namespace `module` to manage your module dependencies:

By default, imports are solved everytime, since solve targets may have sources that change over time. But users can opt in to vendoring their dependencies via the `hlb module` subcommands.

### Module vendoring

```sh
hlb module vendor [--target <ident of an import> ...] <*.hlb>
```

`hlb module vendor` will solve the imports (all imports if no `--target` is specified) and export them to your local system under `./.hlb/modules/**` and basically freeze a snapshot of that hlb library. Since solve targets may have sources like `image "foobar:latest"` that may change when you resolve the import, this is "vendors" the import locally.

Running `hlb module vendor --target <ident of an import>` will update the vendor for a specific import. The path is based on the underlying LLB's vertex digest like `./hlb/modules/sha256/9f/9fabc/module.hlb` so it's stable based on the import fs graph.

### Module print tree

```sh
hlb module tree <*.hlb>
```

Basically prints out a tree of the imports starting from the hlb module you provided. Useful to understand what are your dependencies.

### Cleaning up unused dependencies

```sh
hlb module tidy <*.hlb>
```

When adding and removing imports from a hlb module, the vendor directory will drift and either have unused modules or missing modules. `hlb module tidy` will fill in the gaps and prune unused modules from the directory.

## Import / export declarations

```
import go from fs {
    image "openllb/golang.hlb"
}

export buildHLB
export leftPadNodeModules

fs buildHLB() {
    go.build fs { local "."; } "./cmd/hlb"
}

fs example() {
    image "node:alpine"
    run "npm install" with option {
        dir "/src"
        mount fs { git "https://github.com/left-pad/left-pad.git" "master"; } "/src"
        mount fs { scratch; } "/src/node_modules" as leftPadNodeModules
    }
}
```

Imports declare an identifier that shares the same scope as other function declarations. Then you can access exported functions from the import via `go.<identifier>`. Only functions that are exported may be invoked, so that library authors to define what their public API is.

Exports is a declaration at the top-level scope too, allowing you to export functions and aliases like `as leftPadNodeModules`. Opted for top-level rather than adding modifiers to declarations like `public fs buildHLB` or `as public leftPadNodeModules` so that its clearer what the module import/exports.

## HLB format

`hlb format` is an opinionated way of formatting your `.hlb` files. With the introduction of import/exports, they will be formatted so all the imports declarations are at the top, then exports, then the rest of the function declarations. This way, users can see at a glance, what is imported/exported in this module.

## How libraries are distributed

Since import takes a `fs` function literal, you can import from theoretically any source, be it from a git repo, an OCI image, or a tarball hosted on some HTTP server.

It could even be dynamic:
```
import go from fs {
    scratch
    copy fs {
        image "alpine"
        run "apk add -U openssh-client"
        run "scp root@bastion.com:/home/root/module.hlb /out/module.hlb" with option { ssh; }
    } "/out/module.hlb" "/module.hlb"
}
```

## Import from local hlb modules

Consider the scenario where you import some library and that library is split up into different files, for organizational purposes.

```hlb
# user.hlb
import go from fs {
    image "openllb/golang.hlb"
}
```

```hlb
# golang.hlb
import utils from fs {
    local "./utils.hlb"
}
```

Then when recursively solving the imports, we'll need to:
1. Solve `import go` by unpacking the contents of `openllb/golang.hlb` to local system. (both `golang.hlb` and `utils.hlb`)
2. Parse and check that hlb file.
3. Recursively solve `golang.hlb`'s imports.
4. Transfer `./utils.hlb` back to BuildKit
5. Download `./utils.hlb` again as part of solving an import.

Instead, we could special case local hlb module imports by saying:
```hlb
# golang.hlb
import utils "./utils.hlb`
```

Syntactically different because of missing the `from`, since HLB currently doesn't support polymorphic types so we couldn't overload it there.

## Library wrappers for external frontends

When developing a frontend, you need to also write a function in order to consume it inside HLB. For example, the dockerfile frontend:

```hlb
# dockerfile.hlb
export build

fs build(fs context, fs dockerfile) {
    generate fs {
        image "openllb/dockerfile:experimental" with option { resolve; }
    } with option {
        frontendInput "context" context
        frontendInput "dockerfile" dockerfile
    }
}
```

Then you can also ship another image `openllb/dockerfile.hlb` so that users can consume the dockerfile frontend without copy and pasting the `dockerBuild` function.

```
import dockerfile from fs {
    image "openllb/dockerfile.hlb"
}

fs default() {
    dockerfile.build fs {
        local "." with option { excludePatterns "Dockerfile"; }
    } fs {
        local "." with option { includePatterns "Dockerfile"; }
    }
}
```

**Comment by slushie** [comment]
A few questions about the lockfile:

* What format is the lockfile? Is it intended to be parsed by 3rd party tooling, or should read operations run through the HLB tool?
* If a single module locks its dependencies, how can that lock be shared? Should it be committed alongside the `.hlb` file? 
* Given `import foo from fs { image "myimage"; }` use the `library.hlb` file from _myimage_; would locks also be shipped in that image and used for second-level dependencies?

**Comment by hinshun** [comment]
> * What format is the lockfile? Is it intended to be parsed by 3rd party tooling, or should read operations run through the HLB tool?

The lockfile for hlb is more like vendoring directories but I wasn't sure if the term `vendoring` only exists in golang? When you lock your modules down, you are basically creating a vendored copy into your repository which you will check in. Since HLB dependencies can only be retrieved by solving, you can't regenerate an identical vendor directory unless you control all the remote sources (docker registries, http servers, git repositories, etc).

When vendored, it may look something like:
```sh
❯ tree .hlb
.hlb
└── modules
    └── sha256
        └── 9f
            └── 9ff8d8578c06000a3aa7027e00129524bd3460df2bb6ea3dde0f3bbfd90e4ecc
                └── module.hlb
```

> * If a single module locks its dependencies, how can that lock be shared? Should it be committed alongside the `.hlb` file?

The `.hlb/modules` directory is rooted in the same working directory as the `.hlb` file it was locked for. If that `.hlb` file has siblings, and they have an import with the same `fs` function literal (the vertex hashes to the same digest) then the locked import will be shared.

Yes, the `.hlb/modules` directory should be committed alongside.

> * Given `import foo from fs { image "myimage"; }` use the `module.hlb` file from _myimage_; would locks also be shipped in that image and used for second-level dependencies?

Good point to discuss. In languages like `go` or `node`, the common practice is to not lock library dependencies, but it is an option to the library author. Perhaps the convention here should be the same; i.e. HLB libraries should try not to ship with the `.hlb/modules` directory so that the end-user can decide how to lock them but if they did, their dependencies will be using their own locked modules.

**Comment by slushie** [comment]
I think _vendoring_ is a generic enough term. And I think I conflated the "lock" concept with "lock files" in my mind -- it makes much more sense to me to think of it as a vendor directory. 

Is it possible to store metadata about the actual source for each locked module? In other words, if I later run `hlb module tree` would it show me eg, `import foo from fs { image "bar"; }` as the source of some particular digest hash? Will it also include eg, `resolved image "bar:latest" to "bar@sha256:deadbeef"`?

**Comment by hinshun** [comment]
> I think _vendoring_ is a generic enough term. And I think I conflated the "lock" concept with "lock files" in my mind -- it makes much more sense to me to think of it as a vendor directory.

How about `hlb module vendor` then?

> Is it possible to store metadata about the actual source for each locked module? In other words, if I later run `hlb module tree` would it show me eg, `import foo from fs { image "bar"; }` as the source of some particular digest hash? Will it also include eg, `resolved image "bar:latest" to "bar@sha256:deadbeef"`?

We could show the whole inlined import but it may get pretty long. The current idea is to just show the import name `go` and the digest of the vertex:

```sh
.
└── [go]  sha256:9ff8d8578c06000a3aa7027e00129524bd3460df2bb6ea3dde0f3bbfd90e4ecc
```

Unfortunately, the function literal is opaque so unless I start introspecting the LLB, I wouldn't know what the source is, be it `image` or `git`, etc. Also I'm not sure how to extract cache keys of specific vertices atm.

**Comment by hinshun** [comment]
Changed `hlb mod lock` to `hlb mod vendor`. I think its more intuitive.
